source.go

```diff
@@ -288,12 +323,31 @@ func WithCreateDestPath() llb.CopyOption {
 	})
 }
 
-func SourceIsDir(src Source) bool {
+func (s *Source) IsDir(srcState llb.State, sOpt SourceOpts) bool {
+	return SourceIsDir(*s, srcState, sOpt)
+}
+
+func contextSourceIsDir(srcState llb.State, sOpt SourceOpts) bool {
+	srcFS := sOpt.GetFS(srcState)
+	entries, err := srcFS.ReadDir("/")
+	if err != nil {
+		panic(err)
+	}
+	if len(entries) == 1 {
+		return entries[0].IsDir()
+	} else if len(entries) == 0 {
+		panic("shouldn't be the case") // Maybe this could raise a warning instead
+	}
+	return true
+}
+
+func SourceIsDir(src Source, srcState llb.State, sOpt SourceOpts) bool {
 	switch {
+	case src.Context != nil:
+		return contextSourceIsDir(srcState, sOpt)
```

Comment:
I don't think we should treat context as special.
Also we should only need to inspect the state if `src.Path` is set to anything other than `""` or `"/"`.

Inspecting the state is a pretty expensive call.

Reply:
ok, it is adapted now to only check if the path Is not root and same for the rename.